### ISSUE TITLE
fwupd-refresh.service always fails after hibernate

### ISSUE THREAD

**Issue by fornellas** [fwupd/fwupd#5590]
**Describe the bug**

Every time after waking up from hibernate, `fwupd-refresh.service` fails.

fwupd-refresh.service unit is set to After=network.target, however, in practice, the following happens:

- Desktop is active, NetworkManager is connected.
- Machine is hibernated.
- On wakeup, systemd will start NetworkManager then fwupd-refresh.service.

There's however, a race condition there:

- NetworkManager starts first.
- fwupd-refresh.service is started BEFORE NetworkManager has enough time to reconnect to network.
- fwupd-refresh.service fails due to lack of network.
- NetworkManager connects to network a couple seconds later.

The following journal logs illustrate that:

Feb 27 11:27:27 brown NetworkManager[2048]: <info> [1677497247.0438] device (wlp5s0): state change: unmanaged -> unavailable (reason 'managed', sys-iface-state: 'external')
Feb 27 11:27:27 brown kernel: r8169 0000:06:00.0 enp6s0: Link is Down
Feb 27 11:27:27 brown fwupd[2459754]: 11:27:27:0099 GLib-GObject value "1818326128" of type 'guint' is invalid or out of range for property 'kind' of type 'guint'
Feb 27 11:27:27 brown acpid[2037]: client connected from 4266[0:0]
Feb 27 11:27:27 brown acpid[2037]: 1 client rule loaded
Feb 27 11:27:27 brown NetworkManager[2048]: <info> [1677497247.2354] device (p2p-dev-wlp5s0): state change: unmanaged -> unavailable (reason 'managed', sys-iface-state: 'external')
Feb 27 11:27:27 brown NetworkManager[2048]: <info> [1677497247.2355] manager: NetworkManager state is now CONNECTED_LOCAL
Feb 27 11:27:27 brown systemd[1]: man-db.service: Deactivated successfully.
Feb 27 11:27:27 brown systemd[1]: Finished Daily man-db regeneration.
Feb 27 11:27:27 brown NetworkManager[2048]: <info> [1677497247.2943] device (wlp5s0): supplicant interface state: internal-starting -> disconnected
Feb 27 11:27:27 brown NetworkManager[2048]: <info> [1677497247.2943] device (p2p-dev-wlp5s0): state change: unavailable -> unmanaged (reason 'removed', sys-iface-state: 'removed')
Feb 27 11:27:27 brown NetworkManager[2048]: <info> [1677497247.2947] Wi-Fi P2P device controlled by interface wlp5s0 created
Feb 27 11:27:27 brown NetworkManager[2048]: <info> [1677497247.2949] manager: (p2p-dev-wlp5s0): new 802.11 Wi-Fi P2P device (/org/freedesktop/NetworkManager/Devices/232)
Feb 27 11:27:27 brown NetworkManager[2048]: <info> [1677497247.2951] device (p2p-dev-wlp5s0): state change: unmanaged -> unavailable (reason 'managed', sys-iface-state: 'external')
Feb 27 11:27:27 brown NetworkManager[2048]: <info> [1677497247.2953] device (wlp5s0): state change: unavailable -> disconnected (reason 'supplicant-available', sys-iface-state: 'managed')
Feb 27 11:27:27 brown NetworkManager[2048]: <info> [1677497247.2956] device (p2p-dev-wlp5s0): state change: unavailable -> disconnected (reason 'none', sys-iface-state: 'managed')
Feb 27 11:27:27 brown dbus-daemon[2047]: [system] Successfully activated service 'org.freedesktop.fwupd'
Feb 27 11:27:27 brown systemd[1]: Starting Update APT News...
Feb 27 11:27:27 brown systemd[1]: Started Firmware update daemon.
Feb 27 11:27:27 brown systemd[1]: Starting Update the local ESM caches...
Feb 27 11:27:27 brown fwupdmgr[2459622]: (fwupdmgr:2459622): Fwupd-DEBUG: 11:27:27.743: Emitting ::status-changed() [idle]
Feb 27 11:27:27 brown fwupdmgr[2459622]: Updating lvfs
Feb 27 11:27:27 brown fwupdmgr[2459622]: (fwupdmgr:2459622): Fwupd-DEBUG: 11:27:27.750: downloading https://cdn.fwupd.org/downloads/firmware.xml.gz.jcat
Feb 27 11:27:27 brown fwupdmgr[2459622]: (fwupdmgr:2459622): Fwupd-DEBUG: 11:27:27.751: Emitting ::status-changed() [downloading]
Feb 27 11:27:27 brown fwupdmgr[2459622]: Downloading…: 0%
Feb 27 11:27:27 brown fwupdmgr[2459622]: (fwupdmgr:2459622): Fwupd-DEBUG: 11:27:27.751: Emitting ::status-changed() [idle]
Feb 27 11:27:27 brown systemd[1]: fwupd-refresh.service: Main process exited, code=exited, status=1/FAILURE
Feb 27 11:27:27 brown systemd[1]: fwupd-refresh.service: Failed with result 'exit-code'.
Feb 27 11:27:27 brown systemd[1]: Failed to start Refresh fwupd metadata and update motd.
Feb 27 11:27:27 brown systemd[1]: apt-news.service: Deactivated successfully.
Feb 27 11:27:27 brown systemd[1]: Finished Update APT News.
Feb 27 11:27:28 brown wpa_supplicant[2114]: wlp5s0: Reject scan trigger since one is already pending
Feb 27 11:27:29 brown ModemManager[2147]: <info> [base-manager] couldn't check support for device '/sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:04.0/0000:05:00.0': not supported by any plugin
Feb 27 11:27:29 brown ModemManager[2147]: <info> [base-manager] couldn't check support for device '/sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:05.0/0000:06:00.0': not supported by any plugin
Feb 27 11:27:30 brown wpa_supplicant[2114]: wlp5s0: CTRL-EVENT-REGDOM-CHANGE init=DRIVER type=COUNTRY alpha2=EU
Feb 27 11:27:30 brown NetworkManager[2048]: <info> [1677497250.4717] device (enp6s0): carrier: link connected
Feb 27 11:27:30 brown NetworkManager[2048]: <info> [1677497250.4718] device (enp6s0): state change: unavailable -> disconnected (reason 'carrier-changed', sys-iface-state: 'managed')
Feb 27 11:27:30 brown NetworkManager[2048]: <info> [1677497250.4722] policy: auto-activating connection 'Wired connection 1' (a7e4e975-d8bd-3f26-9ee4-e201b1c20ca8)
Feb 27 11:27:30 brown NetworkManager[2048]: <info> [1677497250.4724] device (enp6s0): Activation: starting connection 'Wired connection 1' (a7e4e975-d8bd-3f26-9ee4-e201b1c20ca8)
Feb 27 11:27:30 brown NetworkManager[2048]: <info> [1677497250.4725] device (enp6s0): state change: disconnected -> prepare (reason 'none', sys-iface-state: 'managed')
Feb 27 11:27:30 brown NetworkManager[2048]: <info> [1677497250.4726] manager: NetworkManager state is now CONNECTING
Feb 27 11:27:30 brown NetworkManager[2048]: <info> [1677497250.4727] device (enp6s0): state change: prepare -> config (reason 'none', sys-iface-state: 'managed')
Feb 27 11:27:30 brown NetworkManager[2048]: <info> [1677497250.4729] device (enp6s0): state change: config -> ip-config (reason 'none', sys-iface-state: 'managed')
Feb 27 11:27:30 brown NetworkManager[2048]: <info> [1677497250.4731] dhcp4 (enp6s0): activation: beginning transaction (timeout in 45 seconds)

The service will be in a failed state, until fwupd-refresh.timer kicks in, which can take up to 12h to happen.

As I have alerts on my machine to detect failed systemd units, this ends up triggering every time the machine is hibernated. Arguably, the alert should be bound by the 12h fwupd-refresh.timer (and only trigger after 12h of being continuously in a fail state, but it seems that addressing this race condition, would still be valid.

One possibility, is to leverage systemd [Service] Restart, and add some retries there, with a couple seconds interval. This should be enough to give time for NetworkManager to get network access working, so fwupd-refresh.service can succeed. Or, we may just add a sleep 60 before /usr/bin/fwupdmgr refresh, same effect.

*Originally reported at https://bugs.launchpad.net/ubuntu/+source/fwupd/+bug/2008699.*

**Steps to Reproduce**
Steps to reproduce the behavior.

- Hibernate the machine.
- Wake up.
- `systemctl status fwupd-refresh.service` should show the failed service; `journalctl` should show the race condition with NetworkManager.

**Expected behavior**
For the fwupd-refresh.service to not always fail on wake up.

**fwupd version information**
Please provide the version of the daemon and client.

```shell
fwupdmgr --version
```

```
runtime   org.freedesktop.fwupd         1.7.9
runtime   com.dell.libsmbios            2.4
compile   org.freedesktop.gusb          0.3.10
compile   com.hughsie.libjcat           0.1.9
runtime   org.kernel                    5.15.0-60-generic
compile   org.freedesktop.fwupd         1.7.9
runtime   org.freedesktop.gusb          0.3.10
```

Please note how you installed it (`apt`, `dnf`, `pacman`, source, etc):

```
dpkg -l | grep fwupd
ii  fwupd                                         1.7.9-1~22.04.1                                    amd64        Firmware update daemon
ii  fwupd-signed                                  1.51~22.04.1+1.2-3ubuntu0.2                        amd64        Linux Firmware Updater EFI signed binary
```

<summary>**fwupd device information**</summary>

Please provide the output of the fwupd devices recognized in your system.

```shell
fwupdmgr get-devices --show-all-devices
```

```
System Product Name
│
├─980 PRO with Heatsink 1TB:
│     Device ID:          71b677ca0f1bc2c5b804fa1d59e52064ce589293
│     Summary:            NVM Express solid state drive
│     Current version:    5B2QGXA7
│     Vendor:             Samsung (NVME:0x144D)
│     Serial Number:      S6WSNS0RB62434Z
│     GUIDs:              4d7a2791-106b-5e72-9cfb-8ea3d89f5421 ← NVME\VEN_144D&DEV_A80A&SUBSYS_144DA801&REV_00
│                         310f81b5-6fce-501e-acfb-487d10501e78 ← NVME\VEN_144D&DEV_A80A&SUBSYS_144DA801
│                         60c89aac-f321-515b-b419-3cf02aa9d375 ← NVME\VEN_144D&DEV_A80A&REV_00
│                         bec63ed7-a95f-54fe-b8cc-8e9fee64ba5a ← NVME\VEN_144D&DEV_A80A
│                         da46b7ec-02b3-5b63-877b-d9deec48d6c1 ← Samsung 980 PRO with Heatsink 1TB
│     Device Flags:       • Internal device
│                         • Updatable
│                         • System requires external power source
│                         • Needs a reboot after installation
│                         • Device is usable for the duration of the update
│                         • Signed Payload
│   
├─INTEL SSDPEKNW010T8:
│     Device ID:          04e17fcf7d3de91da49a163ffe4907855c3648be
│     Summary:            NVM Express solid state drive
│     Current version:    005C
│     Vendor:             Intel Corporation (NVME:0x8086)
│     Serial Number:      PHNH941500391P0B
│     GUIDs:              c5fe8b70-dc9a-5c3b-9634-659091d29812 ← NVME\VEN_8086&DEV_F1A8&SUBSYS_8086390D&REV_03
│                         1122104f-b10a-5f32-bc13-7a1ac0f52ea2 ← NVME\VEN_8086&DEV_F1A8&SUBSYS_8086390D
│                         c6cd9ab0-8f20-512e-9e1f-1af55b8454b9 ← NVME\VEN_8086&DEV_F1A8&REV_03
│                         82741c78-f5dc-5c23-a152-00de5799edc8 ← NVME\VEN_8086&DEV_F1A8
│                         66f90209-aed2-553c-87ee-d8691bd550a7 ← INTEL SSDPEKNW010T8
│     Device Flags:       • Updatable
│                         • System requires external power source
│                         • Needs a reboot after installation
│   
├─Ryzen 7 3700X 8-Core Processor:
│     Device ID:          4bde70ba4e39b28f9eab1628f9dd6e6244c03027
│     Vendor:             AMD
│     GUIDs:              b9a2dd81-159e-5537-a7db-e7101d164d3f ← cpu
│                         997104da-ec11-5e04-9d94-ac5501fbf609 ← CPUID\PRO_0&FAM_17
│                         b54f3c89-8a41-5b9f-9a5c-0367c19a9451 ← CPUID\PRO_0&FAM_17&MOD_71
│                         d951a23c-a920-50ac-ad89-0660aa0be95c ← CPUID\PRO_0&FAM_17&MOD_71&STP_0
│     Device Flags:       • Internal device
│   
├─SSD PLUS 1000GB:
│     Device ID:          74861bcb9108e6a21cde76f31dad912828a79c03
│     Summary:            ATA drive
│     Current version:    UH5100RL
│     Vendor:             SanDisk (ATA:0x15B7, OUI:001b44)
│     Serial Number:      191113803212
│     GUIDs:              0275015d-3172-5540-b4ab-94ac8b6af563 ← IDE\SanDisk_SSD_PLUS_1000GB_________________UH5100RL
│                         dff4341a-b604-5761-ac67-aa8f1fbc500a ← IDE\0SanDisk_SSD_PLUS_1000GB_________________
│                         6594bbc5-e15c-5fe0-898f-b38ee376c318 ← SanDisk SSD PLUS 1000GB
│     Device Flags:       • Updatable
│                         • System requires external power source
│                         • Needs a reboot after installation
│   
├─TPM:
│     Device ID:          c6a80ac3a22083423992a3cb15018989f37834d6
│     Summary:            TPM 2.0 Device
│     Current version:    3.77.0.5
│     Vendor:             AMD (TPM:AMD)
│     GUIDs:              ff71992e-52f7-5eea-94ef-883e56e034c6 ← system-tpm
│                         9305de1c-1e12-5665-81c4-37f8e51219b8 ← TPM\VEN_AMD&DEV_0001
│                         78a291ae-b499-5b0f-8f1d-74e1fefd0b1c ← TPM\VEN_AMD&MOD_AMD
│                         65a3fced-b423-563f-8098-bf5c329fc063 ← TPM\VEN_AMD&DEV_0001&VER_2.0
│                         5e704f0d-83cb-5364-8384-f46d725a23b8 ← TPM\VEN_AMD&MOD_AMD&VER_2.0
│     Device Flags:       • Internal device
│   
├─TU106 [GeForce RTX 2070]:
│     Device ID:          d05d536fa7bd6811a229e6b6e896497fa3cf1084
│     Current version:    a1
│     Vendor:             NVIDIA Corporation (PCI:0x10DE)
│     GUIDs:              1ef1e9a8-4b03-563a-8b20-9c19914d3045 ← PCI\VEN_10DE&DEV_1F02&SUBSYS_14583FDA&REV_A1
│                         40c2dde1-5e44-5caf-b5ab-ea154a3b2390 ← PCI\VEN_10DE&DEV_1F02&SUBSYS_14583FDA
│                         572a863d-55f3-519e-9fed-e541a64bd625 ← PCI\VEN_10DE&DEV_1F02&REV_A1
│                         22806d39-0e3e-5b68-a4c8-c9c82c27be8e ← PCI\VEN_10DE&DEV_1F02
│     Device Flags:       • Internal device
│                         • Cryptographic hash verification is available
│   
├─Unknown Firmware:
│ │   Device ID:          a45df35ac0e948ee180fe216a5f703f32dda163f
│ │   Summary:            UEFI ESRT device
│ │   Current version:    0
│ │   Minimum Version:    72
│ │   Vendor:             DMI:American Megatrends Inc.
│ │   Update State:       Success
│ │   GUIDs:              494350ff-5845-5f31-3200-000000000000
│ │                       230c8b18-8d9b-53ec-838b-6cfc0383493a ← main-system-firmware
│ │   Device Flags:       • Internal device
│ │                       • Updatable
│ │                       • System requires external power source
│ │                       • Needs a reboot after installation
│ │                       • Cryptographic hash verification is available
│ │                       • Device is usable for the duration of the update
│ │ 
│ └─UEFI dbx:
│       Device ID:        362301da643102b9f38477387e2193e57abaa590
│       Summary:          UEFI revocation database
│       Current version:  267
│       Minimum Version:  267
│       Vendor:           UEFI:Linux Foundation
│       Install Duration: 1 second
│       GUIDs:            fda6234b-adcb-5105-8515-9af647d29775 ← UEFI\CRT_D7F66BE77CEF858C174BF4338A99263C8795B74E02026411F5F532F716AE3263
│                         f8ff0d50-c757-5dc3-951a-39d86e16f419 ← UEFI\CRT_D7F66BE77CEF858C174BF4338A99263C8795B74E02026411F5F532F716AE3263&ARCH_X64
│                         c6682ade-b5ec-57c4-b687-676351208742 ← UEFI\CRT_A1117F516A32CEFCBA3F2D1ACE10A87972FD6BBE8FE0D0B996E09E65D802A503
│                         f8ba2887-9411-5c36-9cee-88995bb39731 ← UEFI\CRT_A1117F516A32CEFCBA3F2D1ACE10A87972FD6BBE8FE0D0B996E09E65D802A503&ARCH_X64
│                         7d5759e5-9aa0-5f0c-abd6-7439bb11b9f6 ← UEFI\CRT_ED1FE72CB9CA31C9AF5B757AFCD733323D675825032E6CED7FE1AE9EB767998C
│                         0c7691e1-b6f2-5d71-bc9c-aabee364c916 ← UEFI\CRT_ED1FE72CB9CA31C9AF5B757AFCD733323D675825032E6CED7FE1AE9EB767998C&ARCH_X64
│       Device Flags:     • Internal device
│                         • Updatable
│                         • Supported on remote server
│                         • Needs a reboot after installation
│                         • Only version upgrades are allowed
│                         • Signed Payload
│     
└─WDC WD10EZEX-75WN4A0:
      Device ID:          e1d6939452f58ba2c46050b3033cd16902a8873d
      Summary:            ATA drive
      Current version:    02.01A02
      Vendor:             Western Digital (ATA:0x101C, OUI:0014ee)
      Serial Number:      WD-WCC6Y7LTT0P9
      GUIDs:              d59658aa-1560-5d20-a469-0191428b86ad ← IDE\WDC_WD10EZEX-75WN4A0____________________02.01A02
                          335c8d1a-0ebb-52ac-826b-176e9dc2ca77 ← IDE\0WDC_WD10EZEX-75WN4A0____________________
                          029f24df-26b8-50fb-a3e3-df3c131c5b84 ← WDC WD10EZEX-75WN4A0
      Device Flags:       • Updatable
                          • System requires external power source
                          • Needs a reboot after installation
   
```

**Additional questions**

- Operating system and version: Ubuntu 22.04.2 LTS
- Have you tried rebooting? n/a
- Is this a regression? Yes. One possibility, is to leverage systemd [Service] Restart, and add some retries there, with a couple seconds interval. This should be enough to give time for NetworkManager to get network access working, so fwupd-refresh.service can succeed. Or, we may just add a sleep 60 before /usr/bin/fwupdmgr refresh, same effect.

**Comment by superm1** [comment]
Can you see if this fixes the issue?  https://github.com/fwupd/fwupd/pull/5589  It aligns the unit with that suggested by the systemd documentation.

**Comment by superm1** [comment]
> It seems that 1.8.12 [added After=network-online.target](https://github.com/fwupd/fwupd/commit/fe5d93f7d2712f3b708e562a5f7e9be897e04d1d).

1.8.5 actually is the first version to have it.

> I'm patching my local install of 1.7.9-1 with that and the extra change from https://github.com/fwupd/fwupd/pull/5589:

OK.

**Comment by fornellas** [comment]
I got zero errors since I patched my 1.7.9-1 service. Closing this bug, as it seems the bug only exist in old versions shipped with Ubuntu. Thanks.
